File: packages/modules/lodash/core/typescript/src/array.ts
/* eslint-disable @typescript-eslint/no-explicit-any */

export function chunk<T>(array: T[], size: number = 1): T[][] {
  const length = array.length;
  if (!length || size < 1) return [];
  const result: T[][] = [];
  let index = 0;
  while (index < length) {
    result.push(array.slice(index, (index += size)));
  }
  return result;
}

export function compact<T>(
  array: (T | false | null | undefined | 0 | '')[]
): T[] {
  return array.filter(Boolean) as T[];
}

export function concat<T>(array: T[], ...values: any[]): any[] {
  const result = [...array];
  for (const value of values) {
    if (Array.isArray(value)) {
      result.push(...value);
    } else {
      result.push(value);
    }
  }
  return result;
}

export function difference<T>(array: T[], ...values: T[][]): T[] {
  const exclude = new Set(values.flat());
  return array.filter((item) => !exclude.has(item));
}

export function differenceBy<T>(array: T[], ...args: any[]): T[] {
  const iteratee =
    typeof args[args.length - 1] === 'function' ? args.pop() : undefined;
  const values = args.flat() as T[];
  const mapped = new Set(values.map(iteratee || ((v) => v as unknown)));
  return array.filter(
    (item) => !mapped.has(iteratee ? iteratee(item) : (item as unknown))
  );
}

export function drop<T>(array: T[], n: number = 1): T[] {
  return array.slice(n);
}

export function dropRight<T>(array: T[], n: number = 1): T[] {
  if (n < 0) return [...array];
  return array.slice(0, Math.max(0, array.length - n));
}

export function fill<T>(
  array: T[],
  value: T,
  start: number = 0,
  end: number = array.length
): T[] {
  const result = [...array];
  for (let i = start; i < end && i < result.length; i++) {
    result[i] = value;
  }
  return result;
}

export function findIndex<T>(
  array: T[],
  predicate: (value: T, index: number, obj: T[]) => unknown,
  fromIndex: number = 0
): number {
  for (let i = fromIndex; i < array.length; i++) {
    if (predicate(array[i]!, i, array)) return i;
  }
  return -1;
}

export function findLastIndex<T>(
  array: T[],
  predicate: (value: T, index: number, obj: T[]) => unknown,
  fromIndex: number = array.length - 1
): number {
  for (let i = fromIndex; i >= 0; i--) {
    if (predicate(array[i]!, i, array)) return i;
  }
  return -1;
}

export function flatten<T>(array: (T | T[])[]): T[] {
  const result: T[] = [];
  for (const item of array) {
    if (Array.isArray(item)) {
      result.push(...item);
    } else {
      result.push(item);
    }
  }
  return result;
}

export function flattenDeep<T>(array: any[]): T[] {
  const result: T[] = [];
  function recurse(items: any[]): void {
    for (const item of items) {
      if (Array.isArray(item)) {
        recurse(item);
      } else {
        result.push(item as T);
      }
    }
  }
  recurse(array);
  return result;
}

export function flattenDepth<T>(array: any[], depth: number = 1): T[] {
  if (depth === 0) return array as T[];
  const result: any[] = [];
  for (const item of array) {
    if (Array.isArray(item) && depth > 0) {
      result.push(...flattenDepth(item, depth - 1));
    } else {
      result.push(item);
    }
  }
  return result as T[];
}

export function fromPairs<T>(pairs: [string, T][]): Record<string, T> {
  const result: Record<string, T> = {};
  for (const [key, value] of pairs) {
    result[key] = value;
  }
  return result;
}

export function head<T>(array: T[]): T | undefined {
  return array[0];
}

export function indexOf<T>(
  array: T[],
  value: T,
  fromIndex: number = 0
): number {
  for (let i = fromIndex; i < array.length; i++) {
    if (array[i] === value) return i;
  }
  return -1;
}

export function initial<T>(array: T[]): T[] {
  return array.slice(0, -1);
}

export function intersection<T>(...arrays: T[][]): T[] {
  if (arrays.length === 0) return [];
  const first = arrays[0]!;
  const rest = arrays.slice(1);
  const restSets = rest.map((arr) => new Set(arr));
  return first.filter((item) => restSets.every((set) => set.has(item)));
}

export function intersectionBy<T>(...args: any[]): T[] {
  const iteratee =
    typeof args[args.length - 1] === 'function' ? args.pop() : undefined;
  const arrays: T[][] = args;
  if (arrays.length === 0) return [];
  const first = arrays[0]!;
  const rest = arrays.slice(1);
  const fn: (value: T) => unknown = iteratee || ((v: T) => v as unknown);
  const restMaps = rest.map((arr) => new Set(arr.map(fn)));
  return first.filter((item) => restMaps.every((set) => set.has(fn(item))));
}

export function join<T>(array: T[], separator: string = ','): string {
  let result = '';
  for (let i = 0; i < array.length; i++) {
    if (i > 0) result += separator;
    result += array[i];
  }
  return result;
}

export function last<T>(array: T[]): T | undefined {
  return array[array.length - 1];
}

export function lastIndexOf<T>(
  array: T[],
  value: T,
  fromIndex: number = array.length - 1
): number {
  for (let i = fromIndex; i >= 0; i--) {
    if (array[i] === value) return i;
  }
  return -1;
}

export function nth<T>(array: T[], n: number = 0): T | undefined {
  return n >= 0 ? array[n] : array[array.length + n];
}

export function pull<T>(array: T[], ...values: T[]): T[] {
  const valueSet = new Set(values);
  const result: T[] = [];
  for (const item of array) {
    if (!valueSet.has(item)) {
      result.push(item);
    }
  }
  return result;
}

export function pullAll<T>(array: T[], values: T[]): T[] {
  return pull(array, ...values);
}

export function pullAllBy<T>(
  array: T[],
  values: T[],
  iteratee?: (value: T) => unknown
): T[] {
  const mappedValues = new Set(values.map((v) => (iteratee ? iteratee(v) : v)));
  return array.filter(
    (item) => !mappedValues.has(iteratee ? iteratee(item) : item)
  );
}

export function remove<T>(
  array: T[],
  predicate: (value: T, index: number, obj: T[]) => boolean
): T[] {
  const removed: T[] = [];
  const kept: T[] = [];
  for (let i = 0; i < array.length; i++) {
    const item = array[i]!;
    if (predicate(item, i, array)) {
      removed.push(item);
    } else {
      kept.push(item);
    }
  }
  return removed;
}

export function reverse<T>(array: T[]): T[] {
  return [...array].reverse();
}

export function slice<T>(
  array: T[],
  start: number = 0,
  end: number = array.length
): T[] {
  return array.slice(start, end);
}

export function sortedIndex<T>(array: T[], value: T): number {
  let low = 0;
  let high = array.length;
  while (low < high) {
    const mid = (low + high) >>> 1;
    if (array[mid]! < value) low = mid + 1;
    else high = mid;
  }
  return low;
}

export function sortedIndexBy<T>(
  array: T[],
  value: T,
  iteratee: (value: T) => number
): number {
  let low = 0;
  let high = array.length;
  const valueMapped = iteratee(value);
  while (low < high) {
    const mid = (low + high) >>> 1;
    if (iteratee(array[mid]!) < valueMapped) low = mid + 1;
    else high = mid;
  }
  return low;
}

export function sortedIndexOf<T>(array: T[], value: T): number {
  return sortedIndex(array, value);
}

export function sortedLastIndex<T>(array: T[], value: T): number {
  let low = 0;
  let high = array.length;
  while (low < high) {
    const mid = (low + high) >>> 1;
    if (array[mid]! > value) high = mid;
    else low = mid + 1;
  }
  return low;
}

export function sortedUniq<T>(array: T[]): T[] {
  const result: T[] = [];
  for (let i = 0; i < array.length; i++) {
    if (i === 0 || array[i] !== array[i - 1]) {
      result.push(array[i]!);
    }
  }
  return result;
}

export function sortedLastIndexBy<T>(
  array: T[],
  value: T,
  iteratee: (value: T) => number
): number {
  let low = 0;
  let high = array.length;
  const valueMapped = iteratee(value);
  while (low < high) {
    const mid = (low + high) >>> 1;
    if (iteratee(array[mid]!) > valueMapped) high = mid;
    else low = mid + 1;
  }
  return low;
}

export function tail<T>(array: T[]): T[] {
  return array.slice(1);
}

export function take<T>(array: T[], n: number = 1): T[] {
  return array.slice(0, n);
}

export function takeRight<T>(array: T[], n: number = 1): T[] {
  return n < 0 ? [] : array.slice(array.length - n);
}

export function takeWhile<T>(
  array: T[],
  predicate: (value: T, index: number, obj: T[]) => unknown
): T[] {
  const result: T[] = [];
  for (let i = 0; i < array.length; i++) {
    if (!predicate(array[i]!, i, array)) break;
    result.push(array[i]!);
  }
  return result;
}

export function takeRightWhile<T>(
  array: T[],
  predicate: (value: T, index: number, obj: T[]) => unknown
): T[] {
  const result: T[] = [];
  for (let i = array.length - 1; i >= 0; i--) {
    if (!predicate(array[i]!, i, array)) break;
    result.unshift(array[i]!);
  }
  return result;
}

export function union<T>(...arrays: T[][]): T[] {
  const result: T[] = [];
  const seen = new Set<T>();
  for (const arr of arrays) {
    for (const item of arr) {
      if (!seen.has(item)) {
        seen.add(item);
        result.push(item);
      }
    }
  }
  return result;
}

export function unionBy<T>(...args: any[]): T[] {
  const iteratee =
    typeof args[args.length - 1] === 'function' ? args.pop() : undefined;
  const arrays: T[][] = args;
  const result: T[] = [];
  const seen = new Set<unknown>();
  const fn = iteratee || ((v: T) => v);
  for (const arr of arrays) {
    for (const item of arr) {
      const key = fn(item);
      if (!seen.has(key)) {
        seen.add(key);
        result.push(item);
      }
    }
  }
  return result;
}

export function uniq<T>(array: T[]): T[] {
  return [...new Set(array)];
}

export function uniqBy<T>(array: T[], iteratee: (value: T) => unknown): T[] {
  const result: T[] = [];
  const seen = new Set<unknown>();
  for (const item of array) {
    const key = iteratee(item);
    if (!seen.has(key)) {
      seen.add(key);
      result.push(item);
    }
  }
  return result;
}

export function unzip<T>(arrays: T[][]): T[][] {
  const length = arrays.length ? Math.max(...arrays.map((a) => a.length)) : 0;
  const result: T[][] = [];
  for (let i = 0; i < length; i++) {
    result.push(arrays.map((a) => a[i]!));
  }
  return result;
}

export function unzipWith<T, R>(
  arrays: T[][],
  iteratee: (...values: T[]) => R
): R[] {
  const length = arrays.length ? Math.max(...arrays.map((a) => a.length)) : 0;
  const result: R[] = [];
  for (let i = 0; i < length; i++) {
    result.push(iteratee(...arrays.map((a) => a[i]!)));
  }
  return result;
}

export function without<T>(array: T[], ...values: T[]): T[] {
  return pull(array, ...values);
}

export function xor<T>(...arrays: T[][]): T[] {
  const count = new Map<T, number>();
  for (const arr of arrays) {
    const uniqArr = [...new Set(arr)];
    for (const item of uniqArr) {
      count.set(item, (count.get(item) || 0) + 1);
    }
  }
  const result: T[] = [];
  for (const [item, c] of count) {
    if (c === 1) result.push(item);
  }
  return result;
}

export function xorBy<T>(...args: any[]): T[] {
  const iteratee =
    typeof args[args.length - 1] === 'function' ? args.pop() : undefined;
  const arrays: T[][] = args;
  const count = new Map<unknown, number>();
  const itemMap = new Map<unknown, T>();
  const fn = iteratee || ((v: T) => v as unknown);
  for (const arr of arrays) {
    const uniqArr = [...new Set(arr)];
    for (const item of uniqArr) {
      const key = fn(item);
      count.set(key, (count.get(key) || 0) + 1);
      itemMap.set(key, item);
    }
  }
  const result: T[] = [];
  for (const [key, c] of count) {
    if (c === 1) {
      const item = itemMap.get(key);
      if (item !== undefined) result.push(item);
    }
  }
  return result;
}

export function zip<T>(...arrays: T[][]): T[][] {
  return unzip(arrays);
}

export function zipObject<T>(
  props: (string | number)[],
  values: T[]
): Record<string, T> {
  const result: Record<string, T> = {};
  for (let i = 0; i < props.length; i++) {
    result[String(props[i]!)] = values[i] as T;
  }
  return result;
}

export function zipWith<T, R>(...args: any[]): R[] {
  const iteratee = args.pop();
  const arrays: T[][] = args;
  return unzipWith(arrays, iteratee);
}

File: packages/modules/lodash/core/python/libs/array.py
"""
array
"""

import copy


def chunk(array, size=1):
    """
    chunk
    """
    new_array = []
    for i in range(0, len(array), size):
        new_array.append(array[i : i + size])
    return new_array


def compact(array):
    """
    compact
    """
    new_array = []
    for item in array:
        if item:
            new_array.append(item)
    return new_array


def concat(array, *args):
    """
    concat
    """
    new_array = array
    for arg in args:
        if isinstance(arg, list):
            for item in arg:
                new_array.append(item)
        else:
            new_array.append(arg)
    return new_array


def difference(array, values):
    """
    difference
    """
    if len(array) == 0:
        return []
    _difference = []
    for i in array:
        if i not in values:
            _difference.append(i)
    return _difference


def difference_by(array, values, callback):
    """
    difference_by
    """
    new_values = []
    for value in values:
        new_values.append(callback(value))
    new_array = []
    for item in array:
        if callback(item) not in new_values:
            new_array.append(item)
    return new_array


def drop(array, number=1):
    """
    drop
    """
    return array[number : len(array)]


def drop_right(array, number=1):
    """
    drop_right
    """
    if number >= len(array):
        return []
    return array[0 : len(array) - number]


def drop_right_while(array, callback):
    """
    drop_right_while
    """
    temp_array = copy.deepcopy(array)
    temp_array.reverse()
    for item in temp_array:
        flag = callback(item)
        if flag:
            array.remove(item)


def drop_while(array, callback):
    """
    drop_while
    """
    temp_array = copy.deepcopy(array)
    for item in temp_array:
        flag = callback(item)
        if flag:
            array.remove(item)


def fill(array, value, start=0, end=None):
    """
    fill
    """
    if end is None:
        end = len(array)
    for index, item in enumerate(array):
        if index < start or index > end - 1:
            array[index] = item
        else:
            array[index] = value
    return array


# Linear Search


def find_index(array, callback, from_index=0):
    """
    find_index
    """
    new_array = array[from_index:]
    index = -1
    for idx, item in enumerate(new_array):
        flag = callback(item)
        if flag:
            index = idx
            break
    return index


def find_last_index(array, callback, from_index=None):
    """
    find_last_index
    """
    new_array = array[0:from_index]
    index = -1
    for idx, item in enumerate(new_array):
        flag = callback(item)
        if flag:
            index = idx
    return index


def flatten(array):
    """
    flatten
    """
    new_array = []
    for item in array:
        if isinstance(item, list):
            new_array = concat(new_array, item)
        else:
            new_array.append(item)
    return new_array


def flatten_deep(array):
    """
    flatten_deep
    """
    if array == []:
        return array
    if isinstance(array[0], list):
        return flatten_deep(array[0]) + flatten_deep(array[1:])
    return array[:1] + flatten_deep(array[1:])


def flatten_depth(array, depth=1):
    """
    flatten_depth
    """
    new_array = array
    for _ in range(0, depth):
        new_array = flatten(new_array)
    return new_array


def from_pairs(pairs):
    """
    from_pairs
    """
    object_var = {}
    for pair in pairs:
        key, value = pair
        object_var[key] = value
    return object_var


def head(array):
    """
    head
    """
    if len(array) == 0:
        return
    return array[0]


def index_of(array, value, from_index=0):
    """
    index_of
    """
    if from_index >= len(array):
        return -1
    return array[from_index : len(array)].index(value) + from_index


def initial(array):
    """
    initial
    """
    return array[0 : len(array) - 1]


def intersection(*args):
    """
    intersection
    """
    arrays = list(args)
    total = len(arrays)
    flatten_array = flatten_deep(arrays)
    item_count = {}
    for item in flatten_array:
        if item in item_count:
            item_count[item] += 1
        else:
            item_count[item] = 1
    array = []
    for key, value in item_count.items():
        if value == total:
            array.append(key)
    return array


def join(array, separator):
    """
    join
    """
    return separator.join(array)


def last(array):
    """
    last
    """
    if len(array) == 0:
        return
    return array[len(array) - 1]


def last_index_of(array, value, from_index=None):
    """
    last_index_of
    """
    new_array = array[0:from_index]
    index = -1
    for idx, item in enumerate(new_array):
        if item == value:
            index = idx
    return index


def nth(array, number):
    """
    nth
    """
    return array[number]


def pull(array, *args):
    """
    pull
    """
    values = list(args)
    temp_array = copy.deepcopy(array)
    for item in temp_array:
        if item in values:
            array.remove(item)


def pull_all(array, values):
    """
    pull_all
    """
    temp_array = copy.deepcopy(array)
    for item in temp_array:
        if item in values:
            array.remove(item)


def pull_at(array, indexes):
    """
    pull_at
    """
    temp_array = copy.deepcopy(array)
    removed_array = []
    for index, item in enumerate(temp_array):
        if index in indexes:
            array.remove(item)
            removed_array.append(item)
    return removed_array


def remove(array, callback):
    """
    remove
    """
    new_array = []
    temp_array = copy.deepcopy(array)
    for item in temp_array:
        flag = callback(item)
        if flag:
            array.remove(item)
            new_array.append(item)
    return new_array


# _recursion


def reverse(array):
    """
    reverse
    """
    if len(array) == 0:
        return []
    sub = array[1:]
    reverse_sub = reverse(sub)
    reverse_sub.append(array[0])
    return reverse_sub


def slice(array, start=0, end=None):
    """
    slice
    """
    return array[start:end]


def sorted_index(array, value):
    """
    sorted_index
    """
    index = -1
    for idx, _ in enumerate(array):
        if idx < len(array) - 1:
            first = array[idx]
            second = array[idx + 1]
            if first <= value and value <= second:
                index = idx + 1
                break
        else:
            idx = index + 1
    return index


def sorted_last_index(array, value):
    """
    sorted_last_index
    """
    index = -1
    for idx, _ in enumerate(array):
        if idx < len(array) - 1:
            first = array[idx]
            second = array[idx + 1]
            if first <= value and value <= second:
                index = idx + 1
        else:
            idx = index + 1
    return index


def sorted_uniq(array):
    """
    sorted_uniq
    """
    new_array = []
    for item in array:
        if item not in new_array:
            new_array.append(item)
    return new_array


def tail(array):
    """
    tail
    """
    return array[1 : len(array)]


def take(array, number=1):
    """
    take
    """
    return array[0:number]


def take_right(array, number=1):
    """
    take_right
    """
    length = len(array)
    start = length - number if length - number >= 0 else 0
    return array[start:length]


def take_right_while(array, callback):
    """
    take_right_while
    """
    temp_array = copy.deepcopy(array)
    temp_array.reverse()
    for item in temp_array:
        flag = callback(item)
        if not flag:
            array.remove(item)


def take_while(array, callback):
    """
    take_while
    """
    temp_array = copy.deepcopy(array)
    for item in temp_array:
        flag = callback(item)
        if not flag:
            array.remove(item)


def union(*args):
    """
    union
    """
    arrays = list(args)
    new_array = []
    for array in arrays:
        for item in array:
            if item not in new_array:
                new_array.append(item)
    return new_array


def uniq(array):
    """
    uniq
    """
    return list(set(array))


def unzip(arrays):
    """
    unzip
    """
    total_items_per_array = len(arrays)
    total_arrays = 0
    for array in arrays:
        if total_arrays < len(array):
            total_arrays = len(array)
    new_array = []
    for idx1 in range(0, total_arrays):
        new_array.append([])
        for idx2 in range(0, total_items_per_array):
            value = arrays[idx2][idx1]
            new_array[idx1].append(value)
    return new_array


def unzip_with(arrays, callback):
    """
    unzip_with
    """
    unzipped = unzip(arrays)
    new_array = []
    for array in unzipped:
        value = callback(*array)
        new_array.append(value)
    return new_array


def without(array, *args):
    """
    without
    """
    values = list(args)
    new_array = []
    for item in array:
        if item not in values:
            new_array.append(item)
    return new_array


def xor(*args):
    """
    xor
    """
    arrays = list(args)
    flatten_array = flatten_deep(arrays)
    item_count = {}
    for item in flatten_array:
        if item in item_count:
            item_count[item] += 1
        else:
            item_count[item] = 1
    array = []
    for key, value in item_count.items():
        if value == 1:
            array.append(key)
    return array


def zip(*args):
    """
    zip
    """
    arrays = list(args)
    maximum = 0
    for array in arrays:
        if maximum < len(array):
            maximum = len(array)
    new_arrays = []

    def get_item(array, i):
        try:
            return array[i]
        except Exception:
            return None

    for i in range(0, maximum):
        new_array = []
        for array in arrays:
            item = get_item(array, i)
            new_array.append(item)
        new_arrays.append(new_array)
    return new_arrays


def zip_object(props, values):
    """
    zip_object
    """
    obj = {}
    for index, prop in enumerate(props):
        value = None
        if index < len(values):
            value = values[index]
        obj[prop] = value
    return obj


def _iteratee_value(item, iteratee):
    if iteratee is None:
        return item
    if callable(iteratee):
        return iteratee(item)
    return item[iteratee]


def _is_iteratee(value):
    return callable(value) or isinstance(value, str)


def difference_with(array, values, comparator=None):
    """
    difference_with
    """
    result = []
    for item in array:
        found = False
        for val in values:
            if comparator is None:
                if item == val:
                    found = True
                    break
            elif comparator(item, val):
                found = True
                break
        if not found:
            result.append(item)
    return result


def intersection_by(*arrays):
    """
    intersection_by
    """
    if not arrays:
        return []
    iteratee = arrays[-1] if len(arrays) > 1 and _is_iteratee(arrays[-1]) else None
    if iteratee is not None:
        arrays = arrays[:-1]
    if not arrays:
        return []
    result = []
    seen = set()
    for item in arrays[0]:
        key = _iteratee_value(item, iteratee)
        if key not in seen:
            if all(any(_iteratee_value(x, iteratee) == key for x in arr) for arr in arrays[1:]):
                seen.add(key)
                result.append(item)
    return result


def intersection_with(*arrays):
    """
    intersection_with
    """
    comparator = arrays[-1] if len(arrays) > 1 and callable(arrays[-1]) else None
    if comparator is not None:
        arrays = arrays[:-1]
    if not arrays:
        return []
    result = []
    for item in arrays[0]:
        if all(any(comparator(item, x) for x in arr) for arr in arrays[1:]):
            if not any(comparator(item, r) for r in result):
                result.append(item)
    return result


def pull_all_by(array, values, iteratee=None):
    """
    pull_all_by
    """
    result = []
    for item in array:
        key = _iteratee_value(item, iteratee)
        if not any(_iteratee_value(v, iteratee) == key for v in values):
            result.append(item)
    return result


def pull_all_with(array, values, comparator=None):
    """
    pull_all_with
    """
    result = []
    for item in array:
        if not any(
            item == v if comparator is None else comparator(item, v) for v in values
        ):
            result.append(item)
    return result


def sorted_index_by(array, value, iteratee=None):
    """
    sorted_index_by
    """
    key = _iteratee_value(value, iteratee)
    for i, item in enumerate(array):
        if _iteratee_value(item, iteratee) >= key:
            return i
    return len(array)


def sorted_index_of(array, value):
    """
    sorted_index_of
    """
    for i, item in enumerate(array):
        if item == value:
            return i
        if item > value:
            return -1
    return -1


def sorted_last_index_by(array, value, iteratee=None):
    """
    sorted_last_index_by
    """
    key = _iteratee_value(value, iteratee)
    idx = len(array)
    for i in range(len(array) - 1, -1, -1):
        if _iteratee_value(array[i], iteratee) == key:
            return i + 1
        if _iteratee_value(array[i], iteratee) < key:
            break
    return max(idx, 0)


def sorted_last_index_of(array, value):
    """
    sorted_last_index_of
    """
    idx = -1
    for i, item in enumerate(array):
        if item == value:
            idx = i
        elif item > value:
            break
    return idx


def sorted_uniq_by(array, iteratee=None):
    """
    sorted_uniq_by
    """
    result = []
    seen = set()
    for item in array:
        key = _iteratee_value(item, iteratee)
        if key not in seen:
            seen.add(key)
            result.append(item)
    return result


def union_by(*arrays):
    """
    union_by
    """
    if not arrays:
        return []
    iteratee = arrays[-1] if len(arrays) > 1 and _is_iteratee(arrays[-1]) else None
    if iteratee is not None:
        arrays = arrays[:-1]
    result = []
    seen = set()
    for arr in arrays:
        for item in arr:
            key = _iteratee_value(item, iteratee)
            if key not in seen:
                seen.add(key)
                result.append(item)
    return result


def union_with(*arrays):
    """
    union_with
    """
    comparator = arrays[-1] if len(arrays) > 1 and callable(arrays[-1]) else None
    if comparator is not None:
        arrays = arrays[:-1]
    result = []
    for arr in arrays:
        for item in arr:
            if not any(
                item == r if comparator is None else comparator(item, r) for r in result
            ):
                result.append(item)
    return result


def uniq_by(array, iteratee=None):
    """
    uniq_by
    """
    result = []
    seen = set()
    for item in array:
        key = _iteratee_value(item, iteratee)
        if key not in seen:
            seen.add(key)
            result.append(item)
    return result


def uniq_with(array, comparator=None):
    """
    uniq_with
    """
    result = []
    for item in array:
        if not any(
            item == r if comparator is None else comparator(item, r) for r in result
        ):
            result.append(item)
    return result


def xor_by(*arrays):
    """
    xor_by
    """
    if not arrays:
        return []
    iteratee = arrays[-1] if len(arrays) > 1 and _is_iteratee(arrays[-1]) else None
    if iteratee is not None:
        arrays = arrays[:-1]
    counts = {}
    for arr in arrays:
        seen_in_arr = set()
        for item in arr:
            key = _iteratee_value(item, iteratee)
            if key not in seen_in_arr:
                counts[key] = counts.get(key, 0) + 1
                seen_in_arr.add(key)
    result = []
    seen = set()
    for arr in arrays:
        for item in arr:
            key = _iteratee_value(item, iteratee)
            if counts[key] == 1 and key not in seen:
                seen.add(key)
                result.append(item)
    return result


def xor_with(*arrays):
    """
    xor_with
    """
    comparator = arrays[-1] if len(arrays) > 1 and callable(arrays[-1]) else None
    if comparator is not None:
        arrays = arrays[:-1]
    counts = []
    for arr in arrays:
        arr_group = []
        for item in arr:
            found = False
            for existing, _ in arr_group:
                if comparator is not None and comparator(item, existing):
                    found = True
                    break
                if comparator is None and item == existing:
                    found = True
                    break
            if not found:
                arr_group.append([item, 1])
            else:
                arr_group[-1][1] += 1
        counts.append(arr_group)
    result = []
    for i, arr in enumerate(arrays):
        for item in arr:
            total = 0
            for j, arr_group in enumerate(counts):
                for existing, count in arr_group:
                    match = comparator(item, existing) if comparator else item == existing
                    if match:
                        total += count
                        break
            if total == 1:
                if not any(
                    r == item if comparator is None else comparator(r, item) for r in result
                ):
                    result.append(item)
    return result


def zip_object_deep(props, values):
    """
    zip_object_deep
    """
    obj = {}
    for i, prop in enumerate(props):
        value = None
        if i < len(values):
            value = values[i]
        keys = prop.split(".") if isinstance(prop, str) else [prop]
        current = obj
        for j, key in enumerate(keys):
            if j == len(keys) - 1:
                if isinstance(value, dict) and isinstance(current.get(key), dict):
                    current[key].update(value)
                else:
                    current[key] = value
            else:
                if key not in current:
                    current[key] = {}
                current = current[key]
    return obj


def zip_with(*arrays):
    """
    zip_with
    """
    if not arrays:
        return []
    iteratee = arrays[-1] if len(arrays) > 1 and _is_iteratee(arrays[-1]) else None
    if iteratee is not None:
        arrays = arrays[:-1]
    if not arrays:
        return []
    max_len = max(len(arr) for arr in arrays)
    result = []
    for i in range(max_len):
        values = [arr[i] if i < len(arr) else None for arr in arrays]
        if iteratee is not None:
            result.append(iteratee(*values))
        else:
            result.append(values)
    return result
